plugins/module_utils/common/argspec_validate.py

```diff
@@ -0,0 +1,212 @@
+    if not valid:
+        raise AnsibleActionFail(errors)
+
+"""
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+import json
+import re
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.ansible.utils.plugins.module_utils.common.utils import (
+    dict_merge,
+)
+from ansible.module_utils.six import iteritems, string_types
+from ansible.module_utils._text import to_bytes
+
+try:
+    import yaml
+
+    try:
+        from yaml import CSafeLoader as SafeLoader
+    except ImportError:
+        from yaml import SafeLoader
+    HAS_YAML = True
+except ImportError:
+    HAS_YAML = False
+
+try:
+    from ansible.module_utils.somefile import FutureBaseArgspecValidator
+
+    HAS_ANSIBLE_ARG_SPEC_VALIDATOR = True
+except ImportError:
+    HAS_ANSIBLE_ARG_SPEC_VALIDATOR = False
+
+
+OPTION_METADATA = (
+    "type",
+    "choices",
+    "default",
+    "required",
+    "aliases",
+    "elements",
+    "fallback",
+    "no_log",
+    "apply_defaults",
+    "deprecated_aliases",
+    "removed_in_version",
+)
+OPTION_CONDITIONALS = (
+    "mutually_exclusive",
+    "required_one_of",
+    "required_together",
+    "required_by",
+    "required_if",
+)
+
+VALID_ANSIBLEMODULE_ARGS = (
+    "argument_spec",
+    "bypass_checks",
+    "no_log",
+    "add_file_common_args",
+    "supports_check_mode",
+) + OPTION_CONDITIONALS
+
+BASE_ARG_AVAIL = 2.11
+
+
+class MonkeyModule(AnsibleModule):
+    """A derivative of the AnsibleModule used
+    to just validate the data (task.args) against
+    the schema(argspec)
+    """
+
+    def __init__(self, data, schema, name):
+        self._errors = None
+        self._valid = True
+        self._schema = schema
+        self.name = name
+        self.params = data
+
+    def fail_json(self, msg):
+        """Replace the AnsibleModule fail_json here
+        :param msg: The message for the failure
+        :type msg: str
+        """
+        if self.name:
+            msg = re.sub(
+                r"\(basic\.pyc?\)",
+                "'{name}'".format(name=self.name),
+                msg,
+            )
+        self._valid = False
+        self._errors = msg
+
+    def _load_params(self):
+        """This replaces the AnsibleModule _load_params
+        fn because we already set self.params in init
+        """
+        pass
+
+    def validate(self):
+        """Instantiate the super, validating the schema
+        against the data
+        :return valid: if the data passed
+        :rtype valid: bool
+        :return errors: errors reported during validation
+        :rtype errors: str
+        """
+        super(MonkeyModule, self).__init__(**self._schema)
+        return self._valid, self._errors
+
+
+class AnsibleArgSpecValidator:
+    def __init__(
+        self,
+        data,
+        schema,
+        schema_format,
+        schema_conditionals=None,
+        name=None,
+        other_args=None,
+    ):
+        self._errors = ""
```

Comment:
Can we document the argument list for sake of clarity?

Reply:
good call, yes